quic/s2n-quic-core/src/time/clock.rs

```diff
@@ -48,3 +48,39 @@ impl Clock for NoopClock {
         unsafe { Timestamp::from_duration(Duration::from_micros(1)) }
     }
 }
+
+impl Clock for Timestamp {
+    #[inline]
+    fn get_time(&self) -> Timestamp {
+        *self
+    }
+}
+
+/// A clock that caches the time query for the inner clock
+pub struct Cached<'a, C: Clock> {
+    clock: &'a C,
+    cached_value: core::cell::Cell<Option<Timestamp>>,
+}
+
+impl<'a, C: Clock> Cached<'a, C> {
+    #[inline]
+    pub fn new(clock: &'a C) -> Self {
+        Self {
+            clock,
+            cached_value: Default::default(),
+        }
+    }
+}
+
+impl<'a, C: Clock> Clock for Cached<'a, C> {
+    #[inline]
+    fn get_time(&self) -> Timestamp {
+        if let Some(time) = self.cached_value.get() {
```

Comment:
this cache doesn't need a TTL?

Reply:
Since this borrows an inner clock for `'a`, the lifetime of the borrow is the TTL, if that makes sense. Its intended use is something like this:

```rust
{
    let clock = clock.cached();

    for packet in ack_packets {
        packet.ack_time = clock.get_time();
    }
}

{
    let clock = clock.cached();

    for packet in lost_packets {
        packet.lost_time = clock.get_time();
    }
}
```

Reply:
We've got a few ad-hoc examples of doing this in the codebase than I'm going to clean up after this is merged:

https://github.com/aws/s2n-quic/blob/e2f32e5ed14ae56ac5f11c01039fb745e567bc75/quic/s2n-quic-transport/src/endpoint/mod.rs#L103-L112

Reply:
ah nice. Would it be possible to a add a debug_assertion or something to prevent misuse? Something like if the difference between the cache and now is > x milliseconds, panic?

Reply:
I'm ok to merge this though for now, though if its mainly used in pretty tight loops I think we could start with something like 1 or 2 ms

Reply:
I think that's way too low, especially in debug mode. I was thinking 1s+. But yeah we can figure something else later. I'm not super worried about it being misused since it borrows a clock rather than owning it so it'd really only live as long as a stack frame.